src/components/form/templates/array-field.ts

```diff
@@ -42,43 +114,376 @@ export class ArrayFieldTemplate extends React.Component {
+        const used = new Set<number>();
+
+        for (const index of this.state.order) {
+            const entry = byIndex.get(index);
+            if (!entry) {
+                continue;
+            }
+
+            ordered.push(entry);
+            used.add(index);
+        }
+
+        for (const [index, entry] of byIndex.entries()) {
+            if (!used.has(index)) {
+                ordered.push(entry);
+            }
+        }
 
         return {
-            allowItemRemoval: limeOptions.allowItemRemoval !== false,
-            allowItemReorder: limeOptions.allowItemReorder !== false,
+            ordered,
+            byIndex,
         };
     }
 
-    private handleAddClick(event: MouseEvent) {
-        event.stopPropagation();
-        this.props.onAddClick(event);
+    private readonly handleSortStart = (event: SortableEvent) => {
+        if (!this.canReorderItems()) {
+            return;
+        }
+
+        this.dragSnapshot = [...this.state.order];
+        this.draggedItemIndex = this.getReorderId(event.item);
+
+        if (this.container) {
+            this.container.classList.add(DEFAULT_CONTAINER_CLASS);
+        }
+
+        if (event.item instanceof HTMLElement) {
+            this.applyDropElevation(event.item);
+        }
+    };
+
+    private readonly handleSortEnd = (event: SortableEvent) => {
+        if (!this.canReorderItems()) {
+            return;
+        }
+
+        if (this.container) {
+            this.container.classList.remove(DEFAULT_CONTAINER_CLASS);
+        }
+
+        const snapshot = this.dragSnapshot;
+        const draggedItemIndex = this.draggedItemIndex;
+        this.dragSnapshot = undefined;
+        this.draggedItemIndex = undefined;
+
+        if (event.item instanceof HTMLElement) {
+            this.dropElevationTarget = event.item;
+        }
+
+        const finalOrder = this.readOrderFromDom();
+
+        if (!this.arraysEqual(this.state.order, finalOrder)) {
+            this.setState({ order: finalOrder });
+        }
+
+        if (
+            snapshot === undefined ||
+            draggedItemIndex === undefined ||
+            !(event.item instanceof HTMLElement)
+        ) {
+            this.scheduleDropElevationRemoval();
+            return;
+        }
+
+        const fromPosition = snapshot.indexOf(draggedItemIndex);
+        const toPosition = finalOrder.indexOf(draggedItemIndex);
+
+        if (
+            fromPosition === -1 ||
+            toPosition === -1 ||
+            fromPosition === toPosition
+        ) {
+            this.scheduleDropElevationRemoval();
+            return;
+        }
+
+        const targetItem = (this.props.items ?? []).find((entry) => {
+            return (entry.index ?? -1) === draggedItemIndex;
+        });
+
+        if (!targetItem) {
+            this.scheduleDropElevationRemoval();
+            return;
+        }
+
+        requestAnimationFrame(() => {
+            const reorder = targetItem.onReorderClick(
```

Comment:
I assume the target items are coming from the consumer, so we are trusting them to provide a function here, or we'll crash. The type system _should_ ensure the function exists, but there are many ways for a consumer to bypass that, so I think we should check the function exists before trying to call it.

We can use the `if` on line 274:

```diff
-        if (!targetItem) {
+        if (!targetItem || typeof targetItem.onReorderClick !== 'function') {
             this.scheduleDropElevationRemoval();
             return;
         }
```